File: solutions/python/luhn/1/luhn.py
class Luhn:
    def __init__(self, card_num):
        self.number = card_num.replace(" ", "")

    def valid(self):
        if len(self.number) <= 1 or not self.number.isnumeric():
            return False
        doubled = []
        for i in range(len(self.number) - 2, -1, -2):
            if 2 * int(self.number[i]) > 9:
                doubled.append(2 * int(self.number[i]) - 9)
            else:
                doubled.append(2 * int(self.number[i]))
        for i in range(len(self.number) - 1, -1, -2):
            doubled.append(int(self.number[i]))
        sum_of_num = sum(doubled)
        return sum_of_num % 10 == 0

File: solutions/fsharp/luhn/1/Luhn.fs
module Luhn

open System

let valid (number: string) =
    let cleaned = number.Replace(" ", "")
    if cleaned.Length > 1 && cleaned |> String.forall Char.IsDigit then
        let sum =
            cleaned
            |> Seq.map (string >> int)
            |> Seq.rev
            |> Seq.mapi (fun i d -> if i % 2 = 1 then d * 2 else d)
            |> Seq.map (fun d -> if d > 9 then d - 9 else d)
            |> Seq.sum
        sum % 10 = 0
    else
        false
